File: C++/Grind75/Week-2/67.add-binary.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=67 lang=cpp
 *
 * [67] Add Binary
 */

// @lc code=start
class Solution
{
public:
    string addBinary(string a, string b)
    {
        int i = a.size() - 1, j = b.size() - 1, sum = 0, carry = 0;
        string ans;
        while (i >= 0 || j >= 0)
        {
            sum = carry;

            if (i >= 0)
            {
                sum += a[i--] - '0';
            }
            if (j >= 0)
            {
                sum += b[j--] - '0';
            }

            ans += sum % 2 ? '1' : '0';
            carry = sum > 1 ? 1 : 0;
        }
        if (carry)
            ans += '1';

        reverse(ans.begin(), ans.end());
        return ans;
    }
};
// @lc code=end

File: Java/Grind75/Week2/67.add-binary.java
package Grind75.Week2;

/*
 * @lc app=leetcode id=67 lang=java
 *
 * [67] Add Binary
 */

// @lc code=start
class Solution {
    public String addBinary(String a, String b) {
        StringBuilder ans = new StringBuilder();
        int i = a.length() - 1, j = b.length() - 1, carry = 0;
        while (i >= 0 || j >= 0) {
            int sum = carry;
            if (i >= 0)
                sum += a.charAt(i--) - '0';
            if (j >= 0)
                sum += b.charAt(j--) - '0';
            ans.append(sum % 2);
            carry = sum / 2;
        }
        if (carry > 0)
            ans.append(1);
        return ans.reverse().toString();
    }
}
// @lc code=end